crypto/threads_pthread.c

```diff
@@ -42,12 +45,563 @@
+
+static inline void* fallback_atomic_load_n(void **p)
+{
+    void *ret;
+
+    pthread_mutex_lock(&atomic_sim_lock);
+    ret = *(void **)p;
+    pthread_mutex_unlock(&atomic_sim_lock);
+    return ret;
+}
+
+# define ATOMIC_LOAD_N(p, o) fallback_atomic_load_n((void **)p)
+
+static inline void* fallback_atomic_store_n(void **p, void *v)
+{
+    void *ret;
+
+    pthread_mutex_lock(&atomic_sim_lock);
+    ret = *p;
+    *p = v;
+    pthread_mutex_unlock(&atomic_sim_lock);
+    return ret;
+}
+
+# define ATOMIC_STORE_N(p, v, o) fallback_atomic_store_n((void **)p, (void *)v)
+
+static inline void fallback_atomic_store(void **p, void **v)
+{
+    void *ret;
+    pthread_mutex_lock(&atomic_sim_lock);
+    ret = *p;
+    *p = *v;
+    v = ret;
+    pthread_mutex_unlock(&atomic_sim_lock);
+}
+
+# define ATOMIC_STORE(p, v, o) fallback_atomic_store((void **)p, (void **)v)
+
+static inline void* fallback_atomic_exchange_n(void **p, void *v)
+{
+    void *ret;
+    pthread_mutex_lock(&atomic_sim_lock);
+    ret = *p;
+    *p = v;
+    pthread_mutex_unlock(&atomic_sim_lock);
+    return ret;
+}
+
+#define ATOMIC_EXCHANGE_N(p, v, o) fallback_atomic_exchange_n((void **)p, (void *)v)
+
+static inline uint64_t fallback_atomic_add_fetch(uint64_t *p, uint64_t v)
+{
+    uint64_t ret;
+    pthread_mutex_lock(&atomic_sim_lock);
+    *p += v;
+    ret = *p;
+    pthread_mutex_unlock(&atomic_sim_lock);
+    return ret;
+}
+
+# define ATOMIC_ADD_FETCH(p, v, o) fallback_atomic_add_fetch(p, v)
+
+static inline uint64_t fallback_atomic_fetch_add(uint64_t *p, uint64_t v)
+{
+    uint64_t ret;
+    pthread_mutex_lock(&atomic_sim_lock);
+    ret = *p;
+    *p += v;
+    pthread_mutex_unlock(&atomic_sim_lock);
+    return ret;
+}
+
+# define ATOMIC_FETCH_ADD(p, v, o) fallback_atomic_fetch_add(p, v)
+
+static inline uint64_t fallback_atomic_sub_fetch(uint64_t *p, uint64_t v)
+{
+    uint64_t ret;
+    pthread_mutex_lock(&atomic_sim_lock);
+    *p -= v;
+    ret = *p;
+    pthread_mutex_unlock(&atomic_sim_lock);
+    return ret;
+}
+
+# define ATOMIC_SUB_FETCH(p, v, o) fallback_atomic_sub_fetch(p, v)
+
+static inline uint64_t fallback_atomic_and_fetch(uint64_t *p, uint64_t m)
+{
+    uint64_t ret;
+    pthread_mutex_lock(&atomic_sim_lock);
+    *p &= m;
+    ret = *p;
+    pthread_mutex_unlock(&atomic_sim_lock);
+    return ret;
+}
+
+# define ATOMIC_AND_FETCH(p, v, o) fallback_atomic_and_fetch(p, v)
+
+static inline uint64_t fallback_atomic_or_fetch(uint64_t *p, uint64_t m)
+{
+    uint64_t ret;
+    pthread_mutex_lock(&atomic_sim_lock);
+    *p |= m;
+    ret = *p;
+    pthread_mutex_unlock(&atomic_sim_lock);
+    return ret;
+}
+
+# define ATOMIC_OR_FETCH(p, v, o) fallback_atomic_or_fetch(p, v)
+#endif
+
+static CRYPTO_THREAD_LOCAL rcu_thr_key;
+
+/*
+ * users is broken up into 2 parts
+ * bits 0-15 current readers
+ * bit 32-63 - ID
+ */
+# define READER_SHIFT 0
+# define ID_SHIFT 32
+# define READER_SIZE 16
```

Comment:
There is a `# define READER_SIZE 32` in threads_win.c .
These different values are normal ?
Can you factorize these values into a common header file ?

Reply:
Yes, the values are normal.  A quiescent point contains a 64 bit counter value which contains different fields based on the internal platform implementation.    Currently they could be merged, but I would rather reserve them to be platform dependent should needs change.